active/0020-node-evacuation.md

```diff
@@ -0,0 +1,392 @@
+keep it no matter whether the client is connected or not.
+In 4.3, the session is kept in the memory of the last node to which the client was connected.
+
+So if a node shuts down, sessions of the clients currently or recently connected to it are lost.
+
+Node evacuation should transfer the maximum possible amount of connections and sessions to the other nodes.
+
+After node evacuation is completed a user may shut down the node.
+
+## Concepts
+
+### Takeover process
+
+One of the core concepts concerning EMQX sessions is _session takeover process_.
+
+_Takeover_ process happens when a client with a permanent session reconnects.
+During this process, the session migrates
+* from the node to which the client was previously connected
+* to the node used for serving the new connection.
+
+## Design
+
+We choose to implement node evacuation in two steps.
+
+* Node evacuation without enforced session takeover. In this variant we
+    * discard new connections to the node being evacuated;
+    * disconnect existing node clients;
+    * fix some timeout during which we expect the disconnected clients to reconnect themselves
+    to some other node and takeover their sessions.
+* Node evacuation with enforced session takeover. In this variant, we additionally start takeover processes
+for the sessions kept on the node. Session data is preserved on some other node even if a client does not reconnect itself
+to a new node.
+
+We enable node evacuation capability by introducing a new application.
+
+### Node evacuation **without** enforced session takeover
+
+Node evacuation app has the following components:
+* HTTP API endpoint.
+* Hook logic.
+* CLI.
+
+Node evacuation app on a node has two states:
+* evacuation is off;
+* evacuation is on.
+
+#### Components
+
+##### HTTP API endpoint
+
+HTTP API Endpoint reports the current evacuation status.
+It also sets appropriate HTTP response codes to simplify liveness checks.
+
+Request:
+```
+GET /api/v4/node_evacuation
+```
+
+Responses:
+```
+HTTP/1.1 200 OK
+
+{"status": "disabled"}
+```
+
+```
+HTTP/1.1 503 Service Unavailable
+
+{"status": "enabled"}
+```
+Or:
+```
+HTTP/1.1 503 Service Unavailable
+
+{
+    "status": "enabled",
+    "redirect_to": ["host1:port1", "host2:port2"]
+}
+```
+
+##### CLI
+
+We have the following CLI commands:
+
+```
+emqx_ctl evacuation start --redirect-to host1:port1 host2:port2
+emqx_ctl evacuation start
+emqx_ctl evacuation status
+emqx_ctl evacuation stop
+```
+
+##### Hooks
+
+Evacuation app hooks into the following hook points:
+
+* `client.connect`;
+* `client.connack`;
+* `client.disconnect`.
+
+#### Algorithm
+
+##### Evacuation disabled
+
+This is the default mode when a node starts.
+No hooks are installed by the app. HTTP API endpoint responds with HTTP status `200`.
+
+##### Enabling evacuation
+
+Evacuation is enabled with the CLI command:
+
+```
+emqx_ctl evacuation start --redirect-to host1:port1 host2:port2
```

Comment:
```suggestion
emqx_ctl evacuation start [--redirect-to host1:port1 host2:port2]
```
should be optional, because when the cluster is behind a load balancer, clients have only one endpoint to connect anyway.